### ISSUE TITLE
Send & Recieve fails occasionally

### ISSUE THREAD

**Issue by hahn-kev** [sillsdev/languageforge-lexbox#186]
Unit test logs:
```
System.ApplicationException : transaction abort!
rollback completed
abort: connection ended unexpectedly

hg command was
clone -U  "http://admin:********@hg-staging.languagedepot.org/sena-3" "/tmp/SendReceiveTests/CanCloneSendReceive/Hgweb/admin/6/sena-3" 
Could not get HG VERSION   at Chorus.VcsDrivers.Mercurial.HgRepository.Execute(Boolean failureIsOk, Boolean noChangeIsOk, Int32 secondsBeforeTimeout, String cmd, String[] rest)
   at Chorus.VcsDrivers.Mercurial.HgRepository.CloneFromSource(String sourceLabel, String sourceUri)
   at Chorus.VcsDrivers.Mercurial.HgNormalTransport.Clone()
   at Chorus.VcsDrivers.Mercurial.HgRepository.Clone(RepositoryAddress source, String targetPath, IProgress progress)
   at LfMergeBridge.LanguageForgeMakeCloneActionHandler.LibTriboroughBridgeChorusPlugin.Infrastructure.IBridgeActionTypeHandler.StartWorking(IProgress progress, Dictionary`2 options, String& somethingForClient)
   at LfMergeBridge.LfMergeBridge.Execute(String actionType, IProgress progress, Dictionary`2 options, String& somethingForClient)
   at Testing.Services.SendReceiveService.CloneProject(SendReceiveParams sendReceiveParams, SendReceiveAuth auth) in /github/workspace/backend/Testing/Services/SendReceiveService.cs:line 83
   at Testing.SyncReverseProxy.SendReceiveServiceTests.CanCloneSendReceive(HgProtocol hgProtocol, String user) in /github/workspace/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 127
   at InvokeStub_SendReceiveServiceTests.CanCloneSendReceive(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

dotnet logs:
```
2023-07-13T15:47:42+07:00 info: Yarp.ReverseProxy.Forwarder.HttpForwarder[9]
2023-07-13T15:47:42+07:00       Proxying to http://hg:8088/hg/sena-3?cmd=batch HTTP/2 RequestVersionOrLower no-streaming
2023-07-13T15:47:42+07:00 info: Yarp.ReverseProxy.Forwarder.HttpForwarder[56]
2023-07-13T15:47:42+07:00       Received HTTP/1.1 response 200.
2023-07-13T15:47:43+07:00 info: Yarp.ReverseProxy.Forwarder.HttpForwarder[9]
2023-07-13T15:47:43+07:00       Proxying to http://hg:8088/hg/sena-3?cmd=getbundle HTTP/2 RequestVersionOrLower no-streaming
2023-07-13T15:47:43+07:00 info: Yarp.ReverseProxy.Forwarder.HttpForwarder[56]
2023-07-13T15:47:43+07:00       Received HTTP/1.1 response 200.
2023-07-13T15:47:44+07:00 info: Yarp.ReverseProxy.Forwarder.HttpForwarder[48]
2023-07-13T15:47:44+07:00       ResponseBodyCanceled: Copying the response body was canceled.
2023-07-13T15:47:44+07:00       System.OperationCanceledException: The operation was canceled.
2023-07-13T15:47:44+07:00          at System.Threading.CancellationToken.ThrowOperationCanceledException()
2023-07-13T15:47:44+07:00          at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.TimingPipeFlusher.TimeFlushAsyncAwaited(ValueTask`1 pipeFlushTask, MinDataRate minRate, IHttpOutputAborter outputAborter, CancellationToken cancellationToken)
2023-07-13T15:47:44+07:00          at Yarp.ReverseProxy.Forwarder.StreamCopier.CopyAsync(Stream input, Stream output, Int64 promisedContentLength, StreamCopierTelemetry telemetry, ActivityCancellationTokenSource activityToken, Boolean autoFlush, CancellationToken cancellation)
```

**Comment by hahn-kev** [comment]
Here's a trace for the request that gets terminated:
https://ui.honeycomb.io/sil-language-forge/environments/staging/datasets/lexbox-api/result/8uFdmtrNHWS/trace/uZocmYdfJ3o?fields[]=c_name&fields[]=c_service.name&span=7f0955f44a92b8f7
Logs in hgweb show nothing.

**Comment by rmunn** [comment]
Could this be caused by a timeout? I noticed that the failed test took 44 seconds, which is close enough to 45 seconds that the difference might be rounding error. And our `hgweb/hg.conf` file specifies various timeouts, including `queue-timeout=45`.

I haven't yet been able to reproduce this while running the tests locally, which is making me wonder whether it's simply the fact that GitHub Actions doesn't allocate all that much RAM or CPU time to hosted runners they're providing for free (perfectly reasonably), and so it will run slower than running the same tests on a local machine with lots of RAM and a fast CPU.

I don't have any proof that timeouts are responsible, just a theory so far.

**Comment by rmunn** [comment]
The 45-second timeout that I mentioned was added in commit 300915e. If we find these S/R test failures occurring before that commit, then that would be strong evidence against my theory, as the `queue-timeout` value is unset by default according to the [mod_wsgi docs](https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIDaemonProcess.html). But if these failures only started after commit 300915e. then I'm pretty sure that WSGI timeouts are responsible for the dropped connections we're seeing.

**Comment by rmunn** [comment]
No, that's not it; here's a run where it only took 3 seconds for it to fail: https://github.com/sillsdev/languageforge-lexbox/runs/14975080675

On the other hand, here's a run where we had multiple failures, one in 1 second and one in 43 seconds: https://github.com/sillsdev/languageforge-lexbox/runs/14968636851. The one that failed in just 1 second failed with the error message "Clone response: Clone failure: new repository with no commits. Clone deleted." Which, given that it's trying to clone the sena-3 project, clearly shouldn't have happened: that project should on the staging server should have contents. This has me wondering about timing issues: are we running multiple integration test suites in parallel? Are they interfering with each other somehow?

**Comment by rmunn** [comment]
Mercurial, as of the version we're running, has the following code snippet:

```python
    except httplib.HTTPException:
        raise IOError(None, _('connection ended unexpectedly'))
```

We'll want to change that to be `_('connection ended unexpectedly') + ': ' + repr(e)` (where we have also changed the `except` block to be `except httplib.HTTPException as e` so that we can access `e` inside the block).

Conclusion of our meeting about this: we will add some code to the test runner that will patch Mercurial to change the above code (found in `httppeer.py` to include `repr(e)`). We'll do that patching just before running the unit tests, and the patch should not fail if the code was already patched.

**Comment by rmunn** [comment]
Here's the patch we'll want to apply:

```diff
--- a/mercurial/httppeer.py     Sun Feb 01 18:47:04 2015 -0600
+++ b/mercurial/httppeer.py     Wed Jul 19 16:24:19 2023 +0700
@@ -20,8 +20,8 @@
             while chunk:
                 yield zd.decompress(chunk, 2**18)
                 chunk = zd.unconsumed_tail
-    except httplib.HTTPException:
-        raise IOError(None, _('connection ended unexpectedly'))
+    except httplib.HTTPException as e:
+        raise IOError(None, _('connection ended unexpectedly') + ': ' + repr(e))
     yield zd.flush()
 
 class httppeer(wireproto.wirepeer):
```

**Comment by myieye** [comment]
I don't see any stack traces here that look like this one so... (from [this run](https://github.com/sillsdev/languageforge-lexbox/actions/runs/5795212757/job/15706422818))
```
[xUnit.net 00:01:58.91]     Testing.SyncReverseProxy.SendReceiveServiceTests.CanCloneSendReceive(hgProtocol: Resumable, user: "admin") [FAIL]
  Failed Testing.SyncReverseProxy.SendReceiveServiceTests.CanCloneSendReceive(hgProtocol: Resumable, user: "admin") [56 s]
  Error Message:
   System.IO.IOException : The response ended prematurely, with at least 2737668 additional bytes expected.
  Stack Trace:
     at System.Net.Http.HttpConnection.ContentLengthReadStream.Read(Span`1 buffer)
   at Chorus.Utilities.WebResponseHelper.ReadResponseContent(WebResponse response, Int32 expectedLength)
   at Chorus.VcsDrivers.Mercurial.HgResumeRestApiServer.HandleResponse(HttpWebResponse res)
   at Chorus.VcsDrivers.Mercurial.HgResumeRestApiServer.Execute(String method, HgResumeApiParameters parameters, Byte[] contentToSend, Int32 secondsBeforeTimeout)
   at Chorus.VcsDrivers.Mercurial.HgResumeRestApiServer.Execute(String method, HgResumeApiParameters request, Int32 secondsBeforeTimeout)
   at Chorus.VcsDrivers.Mercurial.HgResumeTransport.PullOneChunk(HgResumeApiParameters request)
   at Chorus.VcsDrivers.Mercurial.HgResumeTransport.Pull(String[] baseRevisions)
   at Chorus.VcsDrivers.Mercurial.HgResumeTransport.Clone()
   at Chorus.VcsDrivers.Mercurial.HgRepository.Clone(RepositoryAddress source, String targetPath, IProgress progress)
   at LfMergeBridge.LanguageForgeMakeCloneActionHandler.LibTriboroughBridgeChorusPlugin.Infrastructure.IBridgeActionTypeHandler.StartWorking(IProgress progress, Dictionary`2 options, String& somethingForClient)
   at LfMergeBridge.LfMergeBridge.Execute(String actionType, IProgress progress, Dictionary`2 options, String& somethingForClient)
   at Testing.Services.SendReceiveService.CloneProject(SendReceiveParams sendReceiveParams, SendReceiveAuth auth) in /github/workspace/backend/Testing/Services/SendReceiveService.cs:line 83
   at Testing.SyncReverseProxy.SendReceiveServiceTests.CanCloneSendReceive(HgProtocol hgProtocol, String user) in /github/workspace/backend/Testing/SyncReverseProxy/SendReceiveServiceTests.cs:line 132
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Comment by hahn-kev** [comment]
Yeah Tim I think that's a separate (though maybe not unrelated issue), that issue is with the Resumable protocol but the other issues here have just been with the standard hg protocol. I'll fork this issue and adjust the name accordingly.

**Comment by hahn-kev** [comment]
I diagnosed some issues yesterday with Tim E. Here's what I believe is a relevant part of a trace ran on lexbox-api
```
5186	10.977982	10.42.3.183	10.42.2.52	TCP	28032	8088 → 43362 [PSH, ACK] Seq=19755493 Ack=2996 Win=64128 Len=27960 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5187	10.977984	10.42.3.183	10.42.2.52	TCP	28032	8088 → 43362 [PSH, ACK] Seq=19783453 Ack=2996 Win=64128 Len=27960 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5188	10.977985	10.42.3.183	10.42.2.52	TCP	8460	8088 → 43362 [PSH, ACK] Seq=19811413 Ack=2996 Win=64128 Len=8388 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5189	10.977987	10.42.3.183	10.42.2.52	TCP	25236	8088 → 43362 [ACK] Seq=19819801 Ack=2996 Win=64128 Len=25164 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5190	10.978085	10.42.3.183	10.42.2.52	TCP	2868	8088 → 43362 [PSH, ACK] Seq=19844965 Ack=2996 Win=64128 Len=2796 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5191	10.978087	10.42.3.183	10.42.2.52	TCP	20230	[TCP Window Full] 8088 → 43362 [PSH, ACK] Seq=19847761 Ack=2996 Win=64128 Len=20158 TSval=3125926126 TSecr=2178017793 [TCP segment of a reassembled PDU]
5192	11.021112	10.42.2.52	10.42.3.183	TCP	72	[TCP ZeroWindow] 43362 → 8088 [ACK] Seq=2996 Ack=19867919 Win=0 Len=0 TSval=2178017836 TSecr=3125926126
5193	11.021270	10.42.3.183	10.42.2.52	TCP	658	[TCP Window Full] [TCP Spurious Retransmission] 8088 → 43362 [ACK] Seq=19867333 Ack=2996 Win=64128 Len=586 TSval=3125926169 TSecr=2178017793
5194	11.021282	10.42.2.52	10.42.3.183	TCP	84	[TCP ZeroWindow] 43362 → 8088 [ACK] Seq=2996 Ack=19867919 Win=0 Len=0 TSval=2178017837 TSecr=3125926169 SLE=19867333 SRE=19867919
5195	11.213094	10.42.2.52	10.42.5.176	TCP	968	[TCP Window Full] HTTP/1.1 200 Script output follows  [TCP segment of a reassembled PDU]
5196	11.213571	10.42.5.176	10.42.2.52	TCP	72	[TCP ZeroWindow] 51208 → 5158 [ACK] Seq=2801 Ack=12806646 Win=0 Len=0 TSval=1014239728 TSecr=501559062
5197	11.245270	10.42.3.183	10.42.2.52	TCP	72	[TCP Keep-Alive] 8088 → 43362 [ACK] Seq=19867918 Ack=2996 Win=64128 Len=0 TSval=3125926393 TSecr=2178017837
5198	11.245296	10.42.2.52	10.42.3.183	TCP	72	[TCP ZeroWindow] 43362 → 8088 [ACK] Seq=2996 Ack=19867919 Win=0 Len=0 TSval=2178018061 TSecr=3125926169
5199	11.449108	10.42.2.52	10.42.5.176	TCP	72	[TCP Keep-Alive] 5158 → 51208 [ACK] Seq=12806645 Ack=2801 Win=64128 Len=0 TSval=501559298 TSecr=1014239728
5200	11.449496	10.42.5.176	10.42.2.52	TCP	72	[TCP ZeroWindow] 51208 → 5158 [ACK] Seq=2801 Ack=12806646 Win=0 Len=0 TSval=1014239963 TSecr=501559062
5201	11.490604	10.42.5.176	10.42.2.52	TCP	72	51208 → 5158 [RST, ACK] Seq=2801 Ack=12806646 Win=3022592 Len=0 TSval=1014240004 TSecr=501559062
```

The first column is the packet #, then the time, then source and destination.
10.42.3.183 is hgweb
10.42.2.52 is lexbox-api
10.42.5.176 is ingress

It looks like hgweb is sending too much data for lexbox to handle and then we're getting spurious retransmission errors finally followed by a TCP reset from ingress, but I'm not sure why. I think I'm going to look into the yarp config and see if I can dig something up.

**Comment by hahn-kev** [comment]
After doing a bunch of research Tim E and I think this is a Dallas specific issue. To test this I deployed to our AWS rancher production instance and was able to do multiple clones of the elawa project without any issues.

One thing I did notice is that the clone from hg was slower than if I executed the same request with the C# http client. Might be worth looking into if a new hg client would have better performance.
